File: handler/novel_workflow.go
package handler

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/tigerowo/freedom/repository"
	"github.com/tigerowo/freedom/service"
)

// === novel-workflow v2：HTTP API ===

// CreateNovelWorkflowRunReq 新建 run 的请求体。
type CreateNovelWorkflowRunReq struct {
	ProjectID   string   `json:"projectId"`
	Mode        string   `json:"mode"`        // auto | manual | quick | custom
	ShotIDs     []string `json:"shotIds"`     // 当前项目分镜 ID 列表（前端从 NovelProject.shotList 读）
	ConfigJSON  string   `json:"configJson"`  // 自定义模式用
}

// CreateNovelWorkflowRun POST /api/v1/novel/workflows
func CreateNovelWorkflowRun(w http.ResponseWriter, r *http.Request) {
	user, ok := service.UserFromContext(r.Context())
	if !ok {
		FailWithStatus(w, http.StatusUnauthorized, "未登录")
		return
	}
	var req CreateNovelWorkflowRunReq
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		FailWithStatus(w, http.StatusBadRequest, "请求体不合法")
		return
	}
	if req.ProjectID == "" {
		FailWithStatus(w, http.StatusBadRequest, "projectId 不能为空")
		return
	}
	run, err := service.CreateNovelWorkflowRun(user.ID, user.GroupID, req.ProjectID, req.Mode, req.ShotIDs, req.ConfigJSON)
	if err != nil {
		FailWithStatus(w, http.StatusInternalServerError, "创建 run 失败: "+err.Error())
		return
	}
	writeJSONWithStatus(w, http.StatusCreated, map[string]any{"code": 0, "data": run})
}

// StartNovelWorkflowRun POST /api/v1/novel/workflows/:id/start
//
// novel-workflow v2 fix: 改用 wrapper 传 id 参数（r.PathValue 在 gin wrapF 下拿不到）
func StartNovelWorkflowRun(w http.ResponseWriter, r *http.Request, id string) {
	user, ok := service.UserFromContext(r.Context())
	if !ok {
		FailWithStatus(w, http.StatusUnauthorized, "未登录")
		return
	}
	if id == "" {
		FailWithStatus(w, http.StatusBadRequest, "id 不能为空")
		return
	}
	run, _ := repository.GetNovelWorkflowRun(id)
	if run == nil || run.UserID != user.ID {
		FailWithStatus(w, http.StatusNotFound, "run 不存在")
		return
	}
	if err := service.StartNovelWorkflowRun(id); err != nil {
		FailWithStatus(w, http.StatusInternalServerError, "启动失败: "+err.Error())
		return
	}
	writeJSON(w, map[string]any{"code": 0, "data": run})
}

// GetNovelWorkflowRun GET /api/v1/novel/workflows/:id
func GetNovelWorkflowRun(w http.ResponseWriter, r *http.Request, id string) {
	user, ok := service.UserFromContext(r.Context())
	if !ok {
		FailWithStatus(w, http.StatusUnauthorized, "未登录")
		return
	}
	run, _ := repository.GetNovelWorkflowRun(id)
	if run == nil || run.UserID != user.ID {
		FailWithStatus(w, http.StatusNotFound, "run 不存在")
		return
	}
	nodes, _ := repository.ListNovelWorkflowNodesByRun(id)
	writeJSON(w, map[string]any{
		"code": 0,
		"data": map[string]any{
			"run":   run,
			"nodes": nodes,
		},
	})
}

// StartNovelWorkflowNode POST /api/v1/novel/workflows/:id/nodes/:nodeId/start
func StartNovelWorkflowNode(w http.ResponseWriter, r *http.Request, runID, nodeID string) {
	user, ok := service.UserFromContext(r.Context())
	if !ok {
		FailWithStatus(w, http.StatusUnauthorized, "未登录")
		return
	}
	run, _ := repository.GetNovelWorkflowRun(runID)
	if run == nil || run.UserID != user.ID {
		FailWithStatus(w, http.StatusNotFound, "run 不存在")
		return
	}
	if err := service.StartNovelWorkflowNode(runID, nodeID); err != nil {
		FailWithStatus(w, http.StatusBadRequest, "启动节点失败: "+err.Error())
		return
	}
	writeJSON(w, map[string]any{"code": 0})
}

// CancelNovelWorkflowNode POST /api/v1/novel/workflows/:id/nodes/:nodeId/cancel
func CancelNovelWorkflowNode(w http.ResponseWriter, r *http.Request, runID, nodeID string) {
	user, ok := service.UserFromContext(r.Context())
	if !ok {
		FailWithStatus(w, http.StatusUnauthorized, "未登录")
		return
	}
	run, _ := repository.GetNovelWorkflowRun(runID)
	if run == nil || run.UserID != user.ID {
		FailWithStatus(w, http.StatusNotFound, "run 不存在")
		return
	}
	if err := service.CancelNovelWorkflowNode(runID, nodeID); err != nil {
		FailWithStatus(w, http.StatusBadRequest, "取消节点失败: "+err.Error())
		return
	}
	writeJSON(w, map[string]any{"code": 0})
}

// RetryNovelWorkflowNode POST /api/v1/novel/workflows/:id/nodes/:nodeId/retry
func RetryNovelWorkflowNode(w http.ResponseWriter, r *http.Request, runID, nodeID string) {
	user, ok := service.UserFromContext(r.Context())
	if !ok {
		FailWithStatus(w, http.StatusUnauthorized, "未登录")
		return
	}
	run, _ := repository.GetNovelWorkflowRun(runID)
	if run == nil || run.UserID != user.ID {
		FailWithStatus(w, http.StatusNotFound, "run 不存在")
		return
	}
	if err := service.RetryNovelWorkflowNode(runID, nodeID); err != nil {
		FailWithStatus(w, http.StatusBadRequest, "重试节点失败: "+err.Error())
		return
	}
	writeJSON(w, map[string]any{"code": 0})
}

// ListNovelWorkflowRuns GET /api/v1/novel/workflows?projectId=...&page=1&pageSize=20
func ListNovelWorkflowRuns(w http.ResponseWriter, r *http.Request) {
	user, ok := service.UserFromContext(r.Context())
	if !ok {
		FailWithStatus(w, http.StatusUnauthorized, "未登录")
		return
	}
	projectID := r.URL.Query().Get("projectId")
	if projectID == "" {
		FailWithStatus(w, http.StatusBadRequest, "projectId 不能为空")
		return
	}
	page, _ := strconv.Atoi(r.URL.Query().Get("page"))
	if page < 1 {
		page = 1
	}
	pageSize, _ := strconv.Atoi(r.URL.Query().Get("pageSize"))
	if pageSize < 1 || pageSize > 100 {
		pageSize = 20
	}
	runs, total, err := repository.ListNovelWorkflowRunsByProject(projectID, pageSize, (page-1)*pageSize)
	if err != nil {
		FailWithStatus(w, http.StatusInternalServerError, "查询失败: "+err.Error())
		return
	}
	// 仅返回当前用户的 run
	filtered := make([]interface{}, 0)
	for _, run := range runs {
		if run.UserID == user.ID {
			filtered = append(filtered, run)
		}
	}
	writeJSON(w, map[string]any{
		"code": 0,
		"data": map[string]any{
			"runs": filtered,
			"total": total,
			"page": page,
			"pageSize": pageSize,
		},
	})
}

File: web/src/services/api/novel_workflow.ts
// novel-workflow v2: 工作流编排层 API client
// 后端路由: /api/v1/novel/workflows/*

import { apiGet, apiPost } from './request';

export type NovelWorkflowMode = 'auto' | 'manual' | 'quick' | 'custom';

export type NovelNodeStatus =
  | '未启动'
  | '排队中'
  | '进行中'
  | '成功'
  | '失败'
  | '跳过'
  | '已取消';

export interface NovelWorkflowRun {
  id: string;
  userId: string;
  userGroupCode: string;
  projectId: string;
  mode: NovelWorkflowMode;
  overallStatus:
    | '未启动'
    | '进行中'
    | '已完成'
    | '部分失败'
    | '已停止';
  configJson: string;
  totalNodes: number;
  successNodes: number;
  failedNodes: number;
  skippedNodes: number;
  canceledNodes: number;
  pendingNodes: number;
  lastOutputUrl: string;
  lastOutputAt: string;
  error: string;
  result: string;
  createdAt: string;
  updatedAt: string;
  startedAt: string;
  completedAt: string;
}

export interface NovelWorkflowNode {
  id: string;
  runId: string;
  userId: string;
  projectId: string;
  nodeId: string;
  layer: 'input' | 'script' | 'asset' | 'shot' | 'post';
  nodeKind: string;
  nodeTitle: string;
  shotId: string;
  shotIndex: number;
  dependsOnJson: string;
  status: NovelNodeStatus;
  progress: number;
  stepMessage: string;
  outputUrl: string;
  genericTaskId: string;
  error: string;
  createdAt: string;
  updatedAt: string;
  startedAt: string;
  completedAt: string;
}

// POST /api/v1/novel/workflows
export async function createNovelWorkflowRun(req: {
  projectId: string;
  mode: NovelWorkflowMode;
  shotIds: string[];
  configJson?: string;
}): Promise<NovelWorkflowRun> {
  const data = await apiPost<NovelWorkflowRun>('/api/v1/novel/workflows', req);
  return data;
}

// GET /api/v1/novel/workflows?projectId=...
export async function listNovelWorkflowRuns(
  projectId: string,
  page = 1,
  pageSize = 20,
): Promise<{
  runs: NovelWorkflowRun[];
  total: number;
  page: number;
  pageSize: number;
}> {
  return apiGet<{
    runs: NovelWorkflowRun[];
    total: number;
    page: number;
    pageSize: number;
  }>('/api/v1/novel/workflows', { projectId, page, pageSize });
}

// GET /api/v1/novel/workflows/:id
export async function getNovelWorkflowRun(runId: string): Promise<{
  run: NovelWorkflowRun;
  nodes: NovelWorkflowNode[];
}> {
  return apiGet<{ run: NovelWorkflowRun; nodes: NovelWorkflowNode[] }>(
    `/api/v1/novel/workflows/${runId}`,
  );
}

// POST /api/v1/novel/workflows/:id/start
export async function startNovelWorkflowRun(runId: string): Promise<NovelWorkflowRun> {
  return apiPost<NovelWorkflowRun>(
    `/api/v1/novel/workflows/${runId}/start`,
  );
}

// POST /api/v1/novel/workflows/:id/nodes/:nodeId/start
export async function startNovelWorkflowNode(
  runId: string,
  nodeId: string,
): Promise<void> {
  await apiPost<void>(
    `/api/v1/novel/workflows/${runId}/nodes/${nodeId}/start`,
  );
}

// POST /api/v1/novel/workflows/:id/nodes/:nodeId/cancel
export async function cancelNovelWorkflowNode(
  runId: string,
  nodeId: string,
): Promise<void> {
  await apiPost<void>(
    `/api/v1/novel/workflows/${runId}/nodes/${nodeId}/cancel`,
  );
}

// POST /api/v1/novel/workflows/:id/nodes/:nodeId/retry
export async function retryNovelWorkflowNode(
  runId: string,
  nodeId: string,
): Promise<void> {
  await apiPost<void>(
    `/api/v1/novel/workflows/${runId}/nodes/${nodeId}/retry`,
  );
}
